### ISSUE TITLE
InstCombine's strcpy(x,x)  -> x transformation removes ASan overlapping-memory checks

### ISSUE THREAD

**Issue by Teemperor** [llvm/llvm-project#55772]
Note: This bug is mainly interesting for [this RFC](https://discourse.llvm.org/t/rfc-safe-optimizations-for-sanitizers/62729). See also issue #55595 and the linked bugs from there.

Example: This triggers an ASan assert in O0 but not with O1.
```cpp
#include <string.h>

int main(int argc, char **argv) {
    char m[3] = {'a', 'b', '\0'};
    return *strcpy(m, m);
}
```
https://godbolt.org/z/h5he4hrdM

The relevant code in InstCombine is:

```cpp
Value *LibCallSimplifier::optimizeStrCpy(CallInst *CI, IRBuilderBase &B) {
  Value *Dst = CI->getArgOperand(0), *Src = CI->getArgOperand(1);
  if (Dst == Src) // strcpy(x,x)  -> x
    return Src;
```

**Comment by nunoplopes** [comment]
TBH, I think these bugs are not useful. I'm sure you can find *hundreds* of bugs of this kind. Are you going to open bug reports for all of them?
First we should settle on a solution. Then you can file reports on missed cases, but for now I think it's not worth it.

**Comment by Teemperor** [comment]
The bugs aren't intended as bug spam (that's why I screen them, link to the root cause to dedup them, put the note at the start, etc.). Sorry if it looked like that.

> First we should settle on a solution. Then you can file reports on missed cases, but for now I think it's not worth it.

Maybe to give some background on how I wanted to approach this:

1. Figure out what the bugs are and if they affect anyone.
2. Start a discussion on how we could fix them and compare required work vs. the value we get (what bugs we fix with what solution vs how big the impact on real users is for each bug fixed).

That seemed reasonable to me. Starting with step 2 might also be an option I guess but the rest of the discussion will end up being "Can we do X?" - "Well, that won't fix issue Y and Z" until we converge on something.  E.g., the current solution people seem to prefer is moving the ASan pass to the front which won't even fix this bug (and other issues).

Anyway, thanks for the input and it's noted.

**Comment by firewave** [comment]
Personally I am interesting in the actual optimization (flag) that is causing these issues not to be detected instead of just knowing the high-level optimization level. This would allow usage of custom optimization flags to keep some performance gains of higher levels and improve detection at the same time.

Though ones that already disappear with `-O1` might warrant a closer look since that is the suggested level for the sanitizers.

**Comment by Teemperor** [comment]
> That's why I don't see value in specific bug reports, because there's no way you can go and fix each of these individually, you need a global solution.

I think there is some misunderstanding here (and I think that's on me because I should have been clearer in the RFC about that): I don't want to actually chase down all these specific optimizations and fix them individually in upstream LLVM. That's why my original pitch was "let's disable a bunch of bad passes in general and make flags like -fno-builtin etc. default with sanitizers".

> For 2), masking bugs is a feature

For some things (e.g. memory leaks) there is an obvious right 'fix', but in general there is now way of knowing whether we mask the bug in the way the user intended the code to actually work. e.g., https://godbolt.org/z/9Yq3xcT3q
